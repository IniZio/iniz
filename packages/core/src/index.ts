export { atom, isAtom } from "./atom";
export { effect, Effect } from "./effect";
export { batch } from "./batch";
