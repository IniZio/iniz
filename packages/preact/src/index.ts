export * from "@iniz/core";
export { useAtom } from "./useAtom";
export { useComputed } from "./useComputed";
export { useSideEffect } from "./useSideEffect";
export { useStore } from "./useStore";
import "./createElement";
