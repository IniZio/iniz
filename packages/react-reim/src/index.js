import React, {PureComponent} from 'react'

import State from './components/state'

export {default as State} from './components/state'

// Syncs props to store
export function pipeTo(store, mutation = (_, p) => p) {
  class Piper extends PureComponent {
    componentDidUpdate() {
      store.setState(mutation, this.props)
    }

    render = () => null
  }
  return Piper
}

export const context = () =>
  ({
    name: 'context',
    call(store) {
      Object.defineProperties(store, {
        __isContext: {
          value: true
        },
        Consumer: {
          value: props => <State store={store} {...props}/>
        },
        get: {
          value: fn => <State store={store}>{fn}</State>
        }
      })
    }
  })

export const createContext = context()

export function connect(store, getter = s => s, setter = () => ({})) {
  const Context = store.__isContext ? store : store.plugin(context())
  return Wrapped => p => (
    <Context.Consumer getter={getter} setter={setter}>
      {
        selected => <Wrapped {...selected} {...p}/>
      }
    </Context.Consumer>
  )
}

export function useReim(store, getter = s => s) {
  const {useState, useEffect} = require('react')

  if (!useState) {
    throw new Error('React@16.7+ is required to use Hooks')
  }

  const [state, setState] = useState(store.snapshot(getter))

  useEffect(() => () => store.unsubscribe(setState))

  store.subscribe(setState, {getter})

  return [state, store.set]
}
