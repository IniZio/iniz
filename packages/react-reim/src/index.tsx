import React, {PureComponent} from 'react'
import reim, {Store, Mutation, Getter} from 'reim'

import State, {StateProps} from './components/state'

export {default as State} from './components/state'

class ReactStore extends Store {
  get __isReactReim() {return true}
  Consumer: (props: StateProps) => React.ReactElement<StateProps>;
  get: (fn: Getter) => React.ReactElement<StateProps>;
}

// Syncs props to store
export function pipeTo(store: Store, mutation: Mutation = (_: any, p: any) => p): React.ComponentClass {
  class Piper extends PureComponent {
    componentDidUpdate() {
      store.set(mutation, this.props)
    }

    render = () => null
  }
  return Piper
}

export const react = () =>
  ({
    name: 'react',
    call(store: Store) {
      Object.defineProperties(store, {
        __isReactReim: {
          value: true
        },
        Consumer: {
          value: (props: any) => <State store={store} {...props}/>
        },
        get: {
          value: (fn: Getter) => <State store={store} getter={fn}/>
        }
      })
    }
  })

export function connect(store: ReactStore, getter: Getter = s => s, setter: Mutation = () => ({})) {
  const Context = store.__isReactReim ? store : store.plugin(react())
  return (Wrapped: React.ComponentClass) => (p: any) => (
    <Context.Consumer getter={getter} setter={setter}>
      {
        selected => <Wrapped {...selected} {...p}/>
      }
    </Context.Consumer>
  )
}

export function useReim(initial: any, getter: Getter = s => s, dependencies: string[] = []) {
  // @ts-ignore
  if (!React.useState) {
    throw new Error('React@16.7-alpha.2 is required to use Hooks')
  }

  const store = initial.__isReim ? initial : reim(initial)

  // @ts-ignore
  const {useState, useEffect, useRef} = React

  const mountRef = useRef()
  const [state, set] = useState(store.snapshot(getter))

  useEffect(() => {
    if (mountRef.current) {
      set(() => store.snapshot(getter))
    } else {
      mountRef.current = true
    }
  }, dependencies)

  useEffect(() => () => store.unsubscribe(set))

  store.subscribe(set, {getter})

  return [state, store.set]
}

export default react
