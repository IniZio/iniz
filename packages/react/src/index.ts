export * from "@reim/core";

export { useAtom } from "./useAtom";
export { useSideEffect } from "./useSideEffect";
export { useCompute } from "./useCompute";