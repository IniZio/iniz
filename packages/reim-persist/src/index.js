import merge from 'lodash/merge'

function canWriteStorage(storage) {
  try {
    storage.setItem('@@', 1);
    storage.removeItem('@@');
    return true;
  } catch (e) {}

  return false;
}

function getState(key, storage, value) {
  try {
    return (value = storage.getItem(key)) && typeof value !== 'undefined'
      ? JSON.parse(value)
      : undefined;
  } catch (err) {}

  return undefined;
}

function setState(key, state, storage) {
  return storage.setItem(key, JSON.stringify(state));
}

export default function persist(options = {}) {
  const storage = options.storage || (window && window.localStorage)
  const key = `reim/${options.key}`

  if (!canWriteStorage(storage)) {
    throw new Error('Invalid storage given')
  }

  return {
    name: 'persist',
    apply(store) {
      const saved = getState(key, storage)
      if (typeof saved === 'object' && saved !== null) {
        store.setState(merge({}, store.state, saved))
      }

      store.subscribe(state => {
        setState(key, state, storage)
      }, {...options.subscriber, immediate: true})

      return store
    }
  }
}
