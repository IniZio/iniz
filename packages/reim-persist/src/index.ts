import {merge} from 'lodash'

function canWriteStorage(storage) {
  try {
    storage.setItem('@@', 1)
    storage.removeItem('@@')
    return true
  } catch (error) {}

  return false
}

function snapshot(key, storage, value?: any) {
  try {
    value = storage.getItem(key)
    return typeof value === 'undefined' ? undefined : JSON.parse(value)
  } catch (error) {}

  return undefined
}

function set(key, state, storage) {
  return storage.setItem(key, JSON.stringify(state))
}

export default function persist(options: any = {}) {
  const storage =
    options.storage ||
    (typeof window === 'undefined' ? null : window && window.localStorage)

  if (!canWriteStorage(storage)) {
    throw new Error('Invalid storage given')
  }

  return (store): void => {
    // console.log('persist store? ', store)

    if (!store.name || store.name.length <= 0) {
      throw new Error(
        'You cannot persist an anonymous store, use `name` option in store'
      )
    }

    const key = `reim/${store.name}`
    const saved = snapshot(key, storage)
    if (typeof saved === 'object' && saved !== null) {
      store.reset(merge({}, store._state, saved))
    }

    store.subscribe(
      state => {
        set(key, state, storage)
      },
      {...options.subscriber, immediate: true}
    )
  }
}
