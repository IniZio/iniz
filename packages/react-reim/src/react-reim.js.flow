// @flow
import {Component} from 'react'

import {Mutation, Getter, Plugin, Store, State} from 'reim'

export type Setter = State | {(state: State):  void | State}

export function pipeTo(store: Store, mutation: Mutation): Component

export function context(): Plugin
export const createContext: Plugin

export const State: Component

export function connect(store: Store, getter: Getter): {(Wrapped: Component): Component}
