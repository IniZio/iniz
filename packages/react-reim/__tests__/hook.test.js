/* eslint react/prop-types: 0 */
/* eslint-disable-next-line import/no-extraneous-dependencies  */
import renderer from 'react-test-renderer'

import reim from '../../reim/src'
import {useReim} from '../src'

import React, {Component, useEffect} from 'react'

test('Hook should return store value on component mount', () => {
  const store = reim({level: 10})

  function TestComponent() {
    const [state, setState] = useReim(store)

    return (
      <div>
        <div id="level">{state.level}</div>
      </div>
    )
  }

  const component = renderer.create(<TestComponent/>)
  store.set(s => {s.level += 2})

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
