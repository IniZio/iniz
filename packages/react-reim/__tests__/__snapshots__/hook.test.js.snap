// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hook should return store value on component mount 1`] = `
<div>
  <div
    id="level"
  >
    12
  </div>
</div>
`;
